lib/icingadb/icingadb.hpp

```diff
@@ -27,6 +31,197 @@
 namespace icinga
 {
 
+/**
+ * RedisKey is the enumeration of all Redis keys used by IcingaDB.
+ *
+ * Each enum value represents a specific Redis key type from which the actual Redis key strings are derived.
+ * For instance, the `Host` enum value corresponds to the Redis key pattern `icinga:host`. These enums help
+ * in organizing and managing the various Redis keys in a transparent and consistent manner and avoid hardcoding
+ * key strings throughout the codebase.
+ *
+ * @ingroup icingadb
+ */
+enum class RedisKey : uint8_t
+{
+	/* Command-related keys */
+	CheckCmdArg,
+	CheckCmdEnvVar,
+	CheckCmdCustomVar,
+
+	EventCmdArg,
+	EventCmdEnvVar,
+	EventCmdCustomVar,
+
+	NotificationCmdArg,
+	NotificationCmdEnvVar,
+	NotificationCmdCustomVar,
+
+	/* Dependency related config */
+	DependencyNode,
+	DependencyEdge,
+	RedundancyGroup,
+
+	/* Hosts & Services */
+	HostCustomVar,
+	HostGroupMember,
+	HostGroupCustomVar,
+
+	ServiceCustomVar,
+	ServiceGroupMember,
+	ServiceGroupCustomVar,
+
+	/* Users & Usergroups */
+	UserCustomVar,
+	UserGroupMember,
+	UserGroupCustomVar,
+
+	/* Notification */
+	NotificationUser,
+	NotificationUserGroup,
+	NotificationRecipient,
+	NotificationCustomVar,
+
+	/* Timeperiods */
+	TimePeriodRange,
+	TimePeriodInclude,
+	TimePeriodExclude,
+	TimePeriodCustomVar,
+
+	/* Downtimes */
+	ScheduledDowntimeCustomVar,
+
+	/* State keys marker and state entries */
+	_state_keys_begin,
+	HostState,
+	ServiceState,
+	RedundancyGroupState,
+	DependencyEdgeState,
+	_state_keys_end,
+};
+
+/**
+ * Dirty bits for config/state changes.
+ *
+ * These are used to mark objects as "dirty" in order to trigger appropriate updates in Redis.
+ * Each bit represents a different type of change that requires a specific action to be taken.
+ *
+ * @ingroup icingadb
+ */
+enum DirtyBits : uint32_t
+{
+	ConfigUpdate  = 1<<0, // Trigger a Redis config update for the object.
+	ConfigDelete  = 1<<1, // Send a deletion command for the object to Redis.
+	VolatileState = 1<<2, // Send a volatile state update to Redis (affects only checkables).
+	RuntimeState  = 1<<3, // Send a runtime state update to Redis (affects only checkables).
+	NextUpdate    = 1<<4, // Update the `icinga:nextupdate:{host,service}` Redis keys (affects only checkables).
+
+	// This bit is set when a child has been registered in a dependency group,
+	// requiring an update of the dependency group's config and state in Redis.
+	DependencyChildRegistered = 1<<5,
+	DependencyGroupState      = 1<<6,
+
+	DeleteSubKeys = 1<<7, // Delete a sub-key in Redis (used for relationships, not associated with an object directly).
```

Comment:
Where's `DeleteSubKeys` actually used? I'd expect some place in `icingadb-worker.cpp` what checks whether this bit is set.